README.md
cytoKernel
====

### `cytoKernel`: An R/Bioconductor package for differential expression using kernel-based score test for high-dimensional biological data.

`cytoKernel` computes the 
feature-wise p values and their corresponding 
adjusted p values in high-dimensional biological experiments.

### Method

> Liu D, Lin X, Ghosh D."Semiparametric regression of
> multi-dimensional genetic pathway data: least-squares 
> kernel machines and linear mixed models". Biometrics. 2007;63(4):1079-1088.
> <doi:10.1111/j.1541-0420.2007.00799.x1028-1039>

> Zhan X, Patterson AD, Ghosh D. "Kernel approaches for   
> differential expression analysis of mass spectrometry-based
> metabolomics data". BMC Bioinformatics. 2015;16:77. Published 2015 Mar 11. 
> <doi:10.1186/s12859-015-0506-3>

> Liu D, Ghosh D, Lin X. "Estimation and testing for the effect > of a genetic pathway on a disease outcome using logistic
> kernel machine regression via logistic mixed models". BMC 
> Bioinf. 2008; 9(1):292.
> <https://doi.org/10.1186/1471-2105-9-292>


### Installing cytoKernel

The R-package **cytoKernel** can be installed from GitHub using the R package
[devtools](https://github.com/hadley/devtools):

Use to install the latest version of **cytoKernel** from GitHub:

    if (!require("devtools")) install.packages("devtools")
    devtools::install_github("Ghoshlab/cytoKernel")

After installation, the package can be loaded into R.
```s
library(cytoKernel)
```
    

### Using cytoKernel

The main function in the **cytoKernel** package is `CytoK()`. The `CytoK()`
function needs two required object and three optional objects: (1)
object: a data frame or a matrix or a Summarized Experiment with one
assay object with observations (e.g., cluster-marker combinations or genes) on the rows.
and samples as the columns (e.g. let’s call it `dataSE`).
(2) group_factor: a binary categorical response variable
that represents the group condition for each sample. For example if the samples represent two different groups or conditions (e.g., before stimulation and after stimulation), provide CytoK() with a phenotype representing which columns in the object are different groups. (e.g. let’s call it `groupSamples`).
(3) lowerRho (optional) a positive value that represents the lower bound of the kernel parameter. Default is 2.
(4) upperRho (optional) a positive value that represents the upper bound of the kernel parameter. Default is 12.
(5) gridRho (optional)* a positive value that represents the number of grid points in the interval of upper and bound of the kernel parameter. Default is 4.
(6) alpha (optional) level of significance to control the False Discovery
rate (FDR). Default is 0.05).
(7) featureVars (optional) Vector of the columns which identify features. If a SummarizedExperiment is used for data, row variables will be used. Default is NULL.

To run the `CytoK()` function,

    CytoKOutput <- CytoK(object = dataSE,
    group_factor =groupSamples,                             lowerRho=2,upperRho=12,gridRho=4,
    alpha = 0.05,featureVars = NULL)

Individual slots can be extracted using accessor methods:

    CytoKFeatures(CytoKOutput) # extracts the data.frame with shrunk effect size, shrunk effect size sd, unadjusted p value and adjusted p value for each feature
    
     CytoKFeaturesOrdered(CytoKOutput) # extracts the data.frame with shrunken effect size, shrunken effect size sd, unadjusted p value and adjusted p value for each feature ordered by unadjusted p value from low to high
     
     CytoKDEfeatures(CytoKOutput) # extracts the percent of differentially expressed features
    
    CytoKData(CytoKOutput) # extracts the original data object
    
    CytoKalpha(CytoKOutput) # extracts the specified level of significance
    
    CytoKFeatureVars(CytoKOutput) # extracts the value of featureVars
    
The heatmap of the expressed matrix of features on rows ordered by the adjusted p values from low to high can be directly
plotted using the `plotCytoK()` function.

     plotCytoK(object = CytoKOutput,
     group_factor =groupSamples,topK=K,...)

For more details, see `vignettes`.

Bug reports
===========

Report bugs as issues on the [GitHub repository new
issue](https://github.com/Ghoshlab/cytoKernel/issues/new)

Contributors
============

-   [Tusharkanti Ghosh](https://github.com/tghosh30)
-   [Victor Lui]()
-   [Pratyaydipta Rudra]()
-   [Souvik Seal]()
-   [Thao Vu]()
-   [Elena Hsieh]()
-   [Debashis Ghosh](https://github.com/ghoshd)



NEWS.md
# cytoKernel 0.99.0

* Submitted `marr 0.99.0` to Bioconductor on October 2, 2020.
* Added a `NEWS.md` file to track changes to the package.


R/CytoK.R
#' @title CytoK
#'
#' @description This function applies a kernel-based score
#' test for identifying differentially expressed features in
#' high-throughput experiments, called the the CytoK procedure. This
#' function also defines the CytoK class and constructor.
#'
#' @param object an object which is a \code{matrix} or
#' \code{data.frame} with features (e.g. cluster-marker
#' combinations or genes) on the rows and samples as the
#' columns. Alternatively, a user can provide a
#' \code{SummarizedExperiment} object and the
#' \code{assay(object)} will be used as input for the
#' CytoK procedure.
#' @param group_factor a group level binary categorical 
#' response associated with each sample or column in the 
#' \code{object}. The order of the \code{group_factor} must 
#' match the order of the columns in \code{object}. 
#' @param lowerRho (Optional) lower bound of the kernel parameter.
#' @param upperRho (Optional) upper bound of the kernel parameter.
#' @param gridRho (Optional) number of grid points in the interval
#' [lowerRho, upperRho].
#' @param alpha (Optional) level of significance to control
#' the False Discovery Rate (FDR).
#' Default is 0.05.
#' @param featureVars (Optional) Vector of the columns which identify
#' features. If a `SummarizedExperiment` is used for `data`,
#' row variables will be used.
#'
#' @return A object of the class \code{CytoK} that
#' contains  a data.frame of the CytoK
#' features in the \code{CytoKFeatures} slot, a data.frame of the CytoK
#' features in the \code{CytoKFeaturesOrdered} slot ordered by
#' adjusted p values from low to high, a numeric value of the CytoK
#' differentially expressed features \code{CytoKDEfeatures} slot, 
#' a data.frame or SummarizedExperiment original data objject 
#' in the \code{CytoKData} slot, a numeric value of the level of
#' significance in the \code{CytoKalpha} slot and (optional)
#' a vector of the columns which identify features in the 
#' \code{CytoKfeatureVars} slot.
#' 
#' @details
#' CytoK (Kernel-based score test in biological feature differential
#' analysis) is a nonlinear approach,
#' which identifies differentially expressed features
#' in high-dimensional biological experiments. This approach can be
#' applied across many different high-dimensional biological data
#' including Flow/Mass Cytometry data and other variety of gene
#' expression data. The CytoK procedure employs a kernel-based
#' score test to identify differentially expressed features.
#' This procedure can be easily applied to a variety of 
#' measurement types since it uses a Gaussian distance based kernel.
#'
#' This function computes the feature-wise p values and their
#' corresponding adjusted p values. Additionally,
#' it also computes the feature-wise shrunk effect sizes and
#' their corresponding shrunk effect size sd's. Further, it 
#' calculates the percent of differentially expressed features. 
#
#' See the vignette for more details.
#'
#' @aliases CytoK
#'
#' @docType methods
#' @name CytoK
#' @importFrom SummarizedExperiment assays
#' @importFrom SummarizedExperiment assay
#' @importFrom SummarizedExperiment colData
#' @importFrom SummarizedExperiment rowData
#' @importFrom dplyr select
#' @importFrom utils combn
#' @importFrom magrittr %>%
#' @examples
#' data <- cbind(matrix(rnorm(1200,mean=2, sd=1.5),
#' nrow=200, ncol=6), matrix(rnorm(1200,mean=5, sd=1.9),
#' nrow=200, ncol=6))
#' data_CytoK <- CytoK(object=data,
#' group_factor = rep(c(0,1), each=6), lowerRho=2,
#' upperRho=12,gridRho=4,alpha = 0.05,
#' featureVars = NULL)
#' data("cytoHDBMW")
#' data_CytoK_HD <- CytoK(object=cytoHDBMW,
#' group_factor = rep(c(0, 1), c(4, 4)), lowerRho=2,
#' upperRho=12,gridRho=4,alpha = 0.05,
#' featureVars = NULL)
#'
#' @references
#' Liu D, Ghosh D, Lin X. Estimation and testing for the 
#' effect of a genetic pathway on a disease outcome using 
#' logistic kernel machine regression via logistic mixed
#' models. BMC Bioinf. 2008; 9(1):292.
#' 
#' Zhan X, Ghosh D. Incorporating auxiliary information for
#' improved prediction using combination of kernel machines.
#' Stat Methodol. 2015; 22:47–57.
#' 
#' Zhan, X., Patterson, A.D. & Ghosh, D.
#' Kernel approaches for differential expression analysis
#' of mass spectrometry-based metabolomics data. 
#' BMC Bioinformatics 16, 77 (2015). 
#' https://doi.org/10.1186/s12859-015-0506-3
#' 
#' Matthew Stephens, False discovery rates: a new deal,
#' Biostatistics, Volume 18, Issue 2, April 2017, 
#' Pages 275–294, https://doi.org/10.1093/biostatistics/kxw041
#' 
#' @rdname CytoK
#' @export
CytoK <- function(object,group_factor,lowerRho=2,
                  upperRho=12,gridRho=4,alpha = 0.05,
                  featureVars = NULL) {
  originalObject <- object
  if (is.data.frame(object)) {
    if(!is.null(featureVars)) {
      featureColumns <- object %>%
        select(featureVars)
      object <- object %>%
        select(-featureVars) %>%
        as.matrix()
    } else {
      object <- as.matrix(object)
      featureColumns <- NULL
    }
    sampleNames <- colnames(object)
  } else if (is(object, "SummarizedExperiment")) {
    featureColumns <- rowData(object)
    sampleNames <- rownames(colData(object))
    object <- assay(object)
  } else if (is(object, "matrix")) {
    if (!is.null(featureVars)) {
      featureColumns <- as.data.frame(object) %>%
        select(featureVars)
      object <- as.data.frame(object) %>%
        select(-featureVars) %>%
        as.matrix()
    } else {
      featureColumns <- NULL
    }
    originalObject <- as.data.frame(object)
    sampleNames <- colnames(as.data.frame(object))
  } else {
    stop("The class of the object must be a matrix,
                    data.frame or SummarizedExperiment")
  }
  
  if(is.null(group_factor)){  
    stop("Must provide group_factor to specify the group 
            level information associated with each sample or 
                 or column in object.")
  }
  
  if(ncol(object) != length(group_factor)){
    stop("Number of columns in object does not match 
                 length of group_factor.")
  }
  
  if(is.factor(group_factor) & length(levels(group_factor)) < 2){
    stop("group_factor is a factor and number of levels in 
            group_factor is less than 2 (levels(group_factor): ",
         levels(group_factor), "). Must provide a factor with 
            exactly 2 levels to use cytoKernel.")
  }
  
  if(is.factor(group_factor) & length(levels(group_factor)) > 2){
    stop("group_factor is a factor and number of levels in 
               group_factor is more than 2 (levels(group_factor): ",
         levels(group_factor), "). Must provide a factor with 
               exactly 2 levels to use cytoKernel.")
  }
  
  if (any(is.na(object))) {
    stop("Object must not contains NAs.")
  }
  CytoKutils <- CytoKProc(object,group_factor,
                          lowerRho, upperRho, gridRho)
  shrunkEffectSizeMean<- CytoKutils$shrunkEffectSizeMean
  shrunkEffectSizeSD<- CytoKutils$shrunkEffectSizeSD
  pValue <- CytoKutils$vec_pValue
  pValueAdj <- CytoKutils$AdjPvalue_features
  DExpfeatures <- (.DEfeatures(pValueAdj,
                               alpha) * 100)/dim(object)[1]
  
  results <- new("CytoK")
  if(is.null(featureColumns)) {
    results@CytoKFeatures <-
      data.frame(EffectSize=shrunkEffectSizeMean,
                 EffectSizeSD=shrunkEffectSizeSD, 
                 pvalue = pValue,padj = pValueAdj)
  } else {
    results@CytoKFeatures <-
      data.frame(featureColumns,
                 EffectSize=shrunkEffectSizeMean,
                 EffectSizeSD=shrunkEffectSizeSD, 
                 pvalue = pValue,padj = pValueAdj)
  }
  if(is.null(featureColumns)) {
    df_null_features <- data.frame(EffectSize=shrunkEffectSizeMean,
                                   EffectSizeSD=shrunkEffectSizeSD, 
                                   pvalue = pValue,padj = pValueAdj)
    results@CytoKFeaturesOrdered <-
      data.frame(df_null_features[order(df_null_features$padj,
                                        decreasing = FALSE),,drop=FALSE])
  } else {
    df_features <- data.frame(featureColumns,
                              EffectSize=shrunkEffectSizeMean,
                              EffectSizeSD=shrunkEffectSizeSD, 
                              pvalue = pValue,padj = pValueAdj)
    results@CytoKFeaturesOrdered <-  
      data.frame(df_features[order(df_features$padj,
                                   decreasing = FALSE),,drop=FALSE])
  }
  results@CytoKDEfeatures <- DExpfeatures
  results@CytoKData <- originalObject
  results@CytoKalpha <- alpha
  results@CytoKFeatureVars <- featureVars
  return(results)
}
